api/debug.go

```diff
@@ -95,3 +131,232 @@ func (d *DebugAPI) TraceBlockByHash(
 
 	return results, nil
 }
+
+func (d *DebugAPI) TraceCall(
+	ctx context.Context,
+	args TransactionArgs,
+	blockNrOrHash rpc.BlockNumberOrHash,
+	config *tracers.TraceCallConfig,
+) (interface{}, error) {
+	if err := rateLimit(ctx, d.ratelimiter, d.logger); err != nil {
+		return nil, err
+	}
+
+	txEncoded, err := encodeTxFromArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	// Default address in case user does not provide one
+	from := d.config.Coinbase
+	if args.From != nil {
+		from = *args.From
+	}
+
+	var traceConfig *tracers.TraceConfig
+	if config != nil {
+		traceConfig = &config.TraceConfig
+	}
+
+	tracer, err := tracerForReceipt(traceConfig, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	height, err := d.resolveBlockNumberOrHash(&blockNrOrHash)
+	if err != nil {
+		return nil, err
+	}
+
+	block, err := d.blocks.GetByHeight(height)
+	if err != nil {
+		return nil, err
+	}
+
+	blockExecutor, err := d.executorAtBlock(block)
+	if err != nil {
+		return nil, err
+	}
+
+	tx := &gethTypes.Transaction{}
+	if err := tx.UnmarshalBinary(txEncoded); err != nil {
+		return nil, err
+	}
+
+	// call tracer
+	head := &gethTypes.Header{
+		Number: big.NewInt(int64(block.Height)),
+		Time:   uint64(time.Now().Unix()),
+	}
+	emulatorConfig := emulator.NewConfig(
+		emulator.WithChainID(d.config.EVMNetworkID),
+		emulator.WithBlockNumber(head.Number),
+		emulator.WithBlockTime(head.Time),
+	)
+	random := block.PrevRandao
+	stateDB := blockExecutor.StateDB
+	tracer.OnTxStart(
+		&tracing.VMContext{
+			Coinbase:    d.config.Coinbase,
+			BlockNumber: head.Number,
+			Time:        head.Time,
+			Random:      &random,
+			GasPrice:    d.config.GasPrice,
+			ChainConfig: emulatorConfig.ChainConfig,
+			StateDB:     stateDB,
+		},
+		tx,
+		from,
+	)
+
+	err = applyStateOverrides(config, stateDB)
+	if err != nil {
+		return nil, err
+	}
+
+	result, err := blockExecutor.Call(from, txEncoded, tracer)
+	if err != nil {
+		return nil, err
+	}
+
+	// call tracer on tx end
+	if tracer.OnTxEnd != nil {
+		tracer.OnTxEnd(result.Receipt(), result.ValidationError)
+	}
+
+	return tracer.GetResult()
+}
+
+func (d *DebugAPI) executorAtBlock(block *models.Block) (*evm.BlockExecutor, error) {
+	blockHeight := block.Height
+	client, err := d.client.GetClientForHeight(blockHeight)
+	if err != nil {
+		return nil, err
+	}
+
+	exeClient, ok := client.(*grpc.Client)
+	if !ok {
+		return nil, fmt.Errorf("could not convert to execution client")
+	}
+
+	ledger, err := evm.NewRemoteLedger(exeClient.ExecutionDataRPCClient(), blockHeight)
+	if err != nil {
+		return nil, fmt.Errorf("could not create remote ledger for height: %d, with: %w", blockHeight, err)
+	}
+
+	return evm.NewBlockExecutor(
+		block,
+		ledger,
+		d.config.FlowNetworkID,
+		d.blocks,
+		d.receipts,
+		d.logger,
+	)
+}
+
+func applyStateOverrides(config *tracers.TraceCallConfig, stateDB types.StateDB) error {
```

Comment:
This is an attempt to implement the original Geth method from here: https://github.com/onflow/go-ethereum/blob/c9c8e533f2f3ae4cad858749edde01b28c789215/internal/ethapi/api.go#L981-L1019

Reply:
This is problematic, blockExecutor.StateDB is shared and it should not. I don't think there is any need for the blockExecutor to have a global StateDB and we can use the one from the emulator when running.

Reply:
Just some context on this. Both `BlockExecutor` and its `StateDB` are short-lived objects. They are created each time `debug_traceCall` is called, and they have no side-effects outside the scope of `debug_traceCall`. Any commitments done through the `StateDB` are not persisted, neither is the `BlockExecutor`.
We cannot use the emulator, because neither `ReadOnlyBlockView` nor `BlockView` expose their `StateDB`.
The `BlockExecutor` operates on a `BlockView` as it needs to execute direct calls, transactions, and dry-run transactions.
We could encapsulate the `StateDB` of `BlockExecutor` and move the `applyStateOverrides` there though.

Reply:
I feel like this should be a method on the executor.

Reply:
> I feel like this should be a method on the executor.

Good point :+1: Updated in https://github.com/onflow/flow-evm-gateway/pull/605/commits/09a1a9661380ca8815232a22fe48773990a04d82